### PR TITLE
Update host card instance preview when edited in outside of app

### PR DESCRIPTION
**Pull request by habdelra**

### CODE REVIEW

**Comment by habdelra** [comment]
One thing I noticed is that the code route preview is more stable (there is less glimmer invalidation). I have a hunch that the resource might need a stale value while the load is in progress. (I saw this in the cards-grid module too when I was making that work for live loading). I'll look into that a bit more...

File: packages/runtime-common/realm.ts [hunk 1]
```diff
@@ -228,21 +236,84 @@ export class Realm {
...
     contents: string,
   ): Promise<{ lastModified: number }> {
+    await this.trackOwnWrite(path);
     let results = await this.#adapter.write(path, contents);
-    await this.#searchIndex.update(this.paths.fileURL(path));
+    await this.#searchIndex.update(this.paths.fileURL(path), {
+      onInvalidation: (invalidatedURLs: URL[]) => {
+        this.sendUpdateMessages({
+          type: 'update',
+          data: { 'index-invalidation': JSON.stringify(invalidatedURLs) },
```
Comment (jurgenwerk) [inline packages/runtime-common/realm.ts]: Would it make sense to put "index-invalidation" value in an enum of event names?

**Review by jurgenwerk — COMMENTED** [review]
I have tested this locally and I think I found a bug. If I go edit an instance (for example I change `firstName` in `packages/drafts-realm/Friend/1.json`), the preview will reload (I see it flicker), but the change will not be reflected (name stays the same). But if I save for the 2nd time, the change will appear.

**Comment by tintinthong** [comment]
> I have tested this locally and I think I found a bug. If I go edit an instance (for example I change `firstName` in `packages/drafts-realm/Friend/1.json`), the preview will reload (I see it flicker), but the change will not be reflected (name stays the same). But if I save for the 2nd time, the change will appear.

I concur with this observation. I tested by opening the monaco file in code route in another tab. The first change isnt picked up but the second one gets picked up. This is despite viewing the 2 incremental indexing events when listening in `curl -H "Accept: text/event-stream" http://localhost:4201/drafts/_message`

**Comment by tintinthong** [comment]
Do I expect this to work for link fields? Like if I edit Person/1.json will Booking.Hosts renders be reflected when looking at isolated view. From what I see the updated data of the links don't render

**Comment by habdelra** [comment]
> Do I expect this to work for link fields? Like if I edit Person/1.json will Booking.Hosts renders be reflected when looking at isolated view. From what I see the updated data of the links don't render

not yet. as I mentioned in the last boxel meeting card instances don't know their dependencies yet. as soon as we implement that, then yes this should work automatically.

**Comment by habdelra** [comment]
> > I have tested this locally and I think I found a bug. If I go edit an instance (for example I change `firstName` in `packages/drafts-realm/Friend/1.json`), the preview will reload (I see it flicker), but the change will not be reflected (name stays the same). But if I save for the 2nd time, the change will appear.
> 
> I concur with this observation. I tested by opening the monaco file in code route in another tab. The first change isnt picked up but the second one gets picked up. This is despite viewing the 2 incremental indexing events when listening in `curl -H "Accept: text/event-stream" http://localhost:4201/drafts/_message`

i think i know what this might be--i had to eliminate an echo that would cause the events of our own writes to re-trigger an index. perhaps that echo cancellation is getting in the way

**Comment by habdelra** [comment]
@tintinthong @jurgenwerk I'm unable to reproduce the issue you guys found, every change that I make (whether in a separate tab via code route editor or VS code) is reflected in the instance preview in code mode and the code route. So i think i'll need to fix this blind. I disabled the echo cacncellation, can you pull the latest and see if you still see the problem? perhaps you can view the `http://localhost:4201/drafts/_message` request in your XHR logs and show me what the event stream looks like?

**Comment by lukemelia** [comment]
I noticed that in Code Mode, if the preview of the card is showing the Embedded or Edit view, a remote edit of the card results in the view reverting back to Isolated. Not sure if this should be considered a bug for this PR.

**Comment by habdelra** [comment]
@lukemelia did you also notice the issues that Justin and Matic ran into?

File: packages/host/app/components/operator-mode/code-mode.gts [hunk 2]
```diff
@@ -28,13 +30,36 @@ interface Signature {
...
+        ) as string[];
+        if (invalidations.includes(card.id)) {
+          this.reloadCard.perform();
+        }
+      }),
+    };
+    registerDestructor(this, () => {
+      if (this.subscription) {
+        this.subscription.unsubscribe();
+      }
```
Comment (lukemelia) [inline packages/host/app/components/operator-mode/code-mode.gts]:
```suggestion
      this.subscription?.unsubscribe();
```

File: packages/host/app/components/editor/go.gts [hunk 3]
```diff
@@ -139,11 +144,34 @@ export default class Go extends Component<Signature> {
...
+        ) as string[];
+        if (invalidations.includes(this.card.id)) {
+          this.loadCard.perform(new URL(this.card.id));
+        }
+      }),
+    };
+    registerDestructor(this, () => {
+      if (this.subscription) {
+        this.subscription.unsubscribe();
+      }
```
Comment (lukemelia) [inline packages/host/app/components/editor/go.gts]:
```suggestion
      this.subscription?.unsubscribe();
```

File: packages/runtime-common/realm.ts [hunk 4]
```diff
@@ -926,9 +997,24 @@ export class Realm {
     );
 
     if (this.listeningClients.length === 0) {
-      await this.#adapter.subscribe((data: Record<string, any>) =>
-        this.sendUpdateMessages({ type: 'update', data }),
-      );
+      await this.#adapter.subscribe((data: Record<string, any>) => {
+        let tracked = this.getTrackedWrite(data);
+        // TODO debugging....
```
Comment (lukemelia) [inline packages/runtime-common/realm.ts]: What's the status of this comment?

**Review by lukemelia — APPROVED** [review]

**Comment by lukemelia** [comment]
> @lukemelia did you also notice the issues that Justin and Matic ran into?

No I couldn't reproduce that.

**Comment by lukemelia** [comment]
> @lukemelia did you also notice the issues that Justin and Matic ran into?

No I couldn't reproduce that.

**Comment by habdelra** [comment]
@lukemelia I reintroduced the echo elimination to this PR. do you wanna pull and try one more time and see if you can see the issues that Matic and Justin ran into? thanks!

**Comment by habdelra** [comment]
for some reason I can't reply to comments in the code:

@lukemelia the status of the "debugging" comment was for Justin and Matic to get back to me on what they were seeing (this is the removal of the echo handling code). I just reverted it since it sounds like you don't see this either. I'm not too sure what is going on...

@jurgenwerk i refactored to type all the SSE events and their data, so it should be really clear what event data is allowed and what is not, and the types can guide us in development

**Comment by lukemelia** [comment]
@habdelra still works for me, although there is a debugger committed at the moment

File: packages/host/app/components/operator-mode/code-mode.gts [hunk 5]
```diff
@@ -28,13 +30,40 @@ interface Signature {
...
+    let url = `${this.cardService.defaultURL}_message`;
+    this.subscription = {
+      url,
+      unsubscribe: this.messageService.subscribe(
+        url,
+        ({ type, data: dataStr }) => {
+          if (type !== 'index') {
+            return;
+          }
+          debugger;
```
Comment (tintinthong) [inline packages/host/app/components/operator-mode/code-mode.gts]: debugger statement

**Review by tintinthong — APPROVED** [review]
@habdelra I pulled recents and did a fresh rebuild and I was unable to recreate the issue previously. Apologies if this was misleading. There is a remaining debugger statement

File: packages/host/app/components/operator-mode/code-mode.gts (hunk 5 above)
Comment (habdelra) [reply packages/host/app/components/operator-mode/code-mode.gts]: oops!

**Pull request merged**